PR extension constructor rebind pprint 4.11

chetmurthy opened the pull request:
This little patch fixes the pretty-printing of "rebind" extension-constructors (and also rebind exceptions) so that it matches the parser.  With tests.  a rebind extension like
```
type t  += A = M.A [@a]
```
was pretty-printed as
```
type t += A[@a] = M.A
```
[obviously wrong, also not accepted by parser]

With tests for extension-constructors and exceptions.

chetmurthy commented:
Arrgh, I think I screwed-up by sending this PR against "trunk" (which apparently is in prep for 4.12?)

If you can let me know which branch I should base my PR against, I'll be happy to redo it -- this is so trivial, it would be a crime to make it more work for the maintainers than a mere "click".

Octachron commented on testsuite/tests/parsetree/source.ml:
```diff
@@ -171,11 +171,13 @@ and[@foo] y = x
 type%foo[@foo] t = int
 and[@foo] t = int
 type%foo[@foo] t += T
+type t += A = M.A[@a]
```
I might be nice to add the more convoluted example:
```ocaml
type t += B = M.A[@b] | C = M.A[@c][@@t]
```

Octachron reviewed (approved):
I find the attribute location in the syntax a bit surprising, but the PR clearly aligns `Pprintast` with the concrete syntax, thanks.
I think that the PR is also fine for 4.11 .

chetmurthy commented:
Let me know if it should go in 4.11, and I can add a Changes file entry.  Thank you for your careful review!

Octachron commented:
Indeed, the change entry can go to the 4.11 section.

Note that I will squash all commits when merging to remove the small review commits from the history (of the repo).

Octachron commented:
@chetmurthy , it looks like I missed this PR in the 4.11 window . I propose to merge this PR immediately. Do you want me to add a Change entry?

chetmurthy commented:
Will also fix this one and update branch, again per your suggestion.

chetmurthy commented:
Florian, could you please remind me, what branch should I base my revised changes upon?

gasche commented:
I believe you should rebase against trunk.

Octachron commented:
Yes, this should be rebased on trunk with a Change entry in the `Working version`/`bugfixes` section.

chetmurthy commented:
pushed a new version of this patch, based on trunk.  Hope it's all in order.

Octachron commented on Changes:
```diff
@@ -359,6 +359,9 @@ Working version
   (Florian Angeletti, review by Gabriel Scherer)
 
+-  #9599: fix pprint of extension constructors (and exceptions) that rebind
+   (Chet Murthy, review by octachron@)
```
`Florian Angeletti` rather than `octachron@`. Also the Change entry is conflicting with the one in the previous PR. Do you want me to (force) push a fixed and rebased version?

chetmurthy replied:
There's no point in you doing this sort of minor work; let me do it and I'll push an updated rebased branch.

chetmurthy replied:
OK, pushed a rebased branch, builds, passes all tests.

chetmurthy committed:
redo: This little patch fixes the pretty-printing of "rebind" extension-constructors (and also rebind exceptions) so that it matches the parser. With tests. a rebind extension like

type t  += A = M.A [@a]
was pretty-printed as

type t += A[@a] = M.A
[obviously wrong, also not accepted by parser]

With tests for extension-constructors and exceptions.

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch pr-extension-constructor-rebind-pprint-4.11
# Changes to be committed:
#	modified:   Changes
#	modified:   parsing/pprintast.ml
#	modified:   testsuite/tests/parsetree/source.ml
#
# Untracked files:
#	Changes.orig
#	parsing/pprintast.ml.orig
#	testsuite/tests/parsetree/source.ml.orig
#	testsuite/tests/parsetree/source.ml.rej
#
